update-engine/src/errors.rs

```diff
@@ -51,3 +51,68 @@ impl<S: StepSpec> From<mpsc::error::SendError<Event<S>>> for ExecutionError<S> {
         Self::EventSendError(value)
     }
 }
+
+/// Returns an error that occurred while converting an event into its generic
+/// form.
+#[derive(Debug)]
+pub struct ConvertGenericError {
+    pub path: VecDeque<ConvertGenericPathElement>,
+    pub error: serde_json::Error,
+}
+
+impl ConvertGenericError {
+    pub(crate) fn new(elem: &'static str, error: serde_json::Error) -> Self {
+        Self {
+            path: VecDeque::from_iter([ConvertGenericPathElement::Path(elem)]),
+            error,
+        }
+    }
+
+    pub(crate) fn parent(mut self, elem: &'static str) -> Self {
+        self.path.push_front(ConvertGenericPathElement::Path(elem));
+        self
+    }
+
+    pub(crate) fn parent_array(
+        mut self,
+        elem: &'static str,
+        index: usize,
+    ) -> Self {
+        self.path
+            .push_front(ConvertGenericPathElement::ArrayIndex(elem, index));
+        self
+    }
+}
+
+impl fmt::Display for ConvertGenericError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("error converting path: ")?;
+        // VecDeque doesn't have join()...
```

Comment:
I don't think this comment is necessary - `join()` would create a temp string that this loop doesn't, right? Either way, the loop seems perfectly reasonable to me.